pkg/bootstrap/bootstrap.go

```diff
@@ -187,54 +244,167 @@ func extractFromDir(dir, prefix string, img v1.Image, imgName string) error {
+// in that directory as HelmCharts. Any manifests that contain a HelmChart are modified to
+// pass through the systemDefaultRegistry setting to both the Helm job and the chart values.
+// NOTE: This will probably fail if any manifest contains multiple documents. This should
+// not matter for any of our packaged components, but may prevent this from working on user manifests.
+func setSystemDefaultRegistry(manifestsDir string, systemDefaultRegistry string) error {
+	serializer := json.NewSerializerWithOptions(json.DefaultMetaFactory, schemes.All, schemes.All, json.SerializerOptions{Yaml: true, Pretty: true, Strict: true})
+
+	files := map[string]os.FileInfo{}
+	if err := filepath.Walk(manifestsDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		switch {
+		case info.IsDir():
+			return nil
+		case strings.HasSuffix(path, ".yml"):
+		case strings.HasSuffix(path, ".yaml"):
+		default:
+			return nil
+		}
+		files[path] = info
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	var errs []error
+	for fileName, info := range files {
+		if err := rewriteChart(fileName, info, systemDefaultRegistry, serializer); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return merr.NewErrors(errs...)
+}
+
+// rewriteChart applies systemDefaultRegistry settings to the file at fileName with associated info.
+// If the file cannot be decoded as a HelmChart, it is silently skipped. Any other IO error is considered
+// a failure.
+func rewriteChart(fileName string, info os.FileInfo, systemDefaultRegistry string, serializer *json.Serializer) error {
+	chartChanged := false
+
+	bytes, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return errors2.Wrapf(err, "Failed to read manifest %q", fileName)
+	}
+
+	// Ignore manifest if it cannot be decoded
+	obj, _, err := serializer.Decode(bytes, nil, nil)
+	if err != nil {
+		logrus.Debugf("Failed to decode manifest %q: %s", fileName, err)
+		return nil
+	}
+
+	// Ignore manifest if it is not a HelmChart
+	chart, ok := obj.(*helmv1.HelmChart)
+	if !ok {
+		logrus.Debugf("Manifest %q is not a HelmChart", fileName)
+		return nil
+	}
+
+	// Generally we should avoid using Set on HelmCharts since it cannot be overridden by HelmChartConfig,
+	// but in this case we need to do it in order to avoid potentially mangling the ValuesContent field by
+	// blindly appending content to it in order to set the global.systemDefaultRegistry value.
+	if chart.Spec.Set == nil {
+		chart.Spec.Set = map[string]intstr.IntOrString{}
+	}
+	if chart.Spec.Set["global.systemDefaultRegistry"].StrVal != systemDefaultRegistry {
+		chart.Spec.Set["global.systemDefaultRegistry"] = intstr.FromString(systemDefaultRegistry)
+		chartChanged = true
+	}
+
+	jobImage := helm.DefaultJobImage
+	if systemDefaultRegistry != "" {
+		jobImage = systemDefaultRegistry + "/" + helm.DefaultJobImage
+	}
+
+	if chart.Spec.JobImage != jobImage {
+		chart.Spec.JobImage = jobImage
```

Comment:
This assumes the chart-defined `JobImage` is the default and clobbers anything the user chart may have specified.

Reply:
Correct. This is necessary to ensure that the current flag value is applied to the files on disk so that they can get picked up by the deploy controller.

Do you think we should actually rewrite the image string to apply the registry setting, instead of just replacing it? Or only change it if it's empty or ends with the default value? 

cc: @cjellick